### PR TITLE
crush: add multistep retry rules

### PR DESCRIPTION
**Pull request by athanatos**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x ] New feature (ticket optional)
  - [ x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
Added thrash-erasure-code big tests pass https://pulpito.ceph.com/sjust-2024-01-04_18:17:17-rados:thrash-erasure-code-big-wip-sjust-msr-testing-2024-01-03-distro-default-smithi/

**Comment by athanatos** [comment]
Checking/fixing make check failures -- related to tool output differences

File: doc/rados/operations/crush-map-edits.rst [hunk 1]
```diff
@@ -430,6 +430,18 @@ A rule takes the following form::
         step emit
     }
 
+CRUSH MSR rules are a distinct type of CRUSH rule which supports retrying steps
```
Comment (ljflores) [inline doc/rados/operations/crush-map-edits.rst:433]:
```suggestion
CRUSH MSR (multi-step retry) rules are a distinct type of CRUSH rule which supports retrying steps
```

File: src/crush/mapper.c [hunk 2]
```diff
@@ -1081,3 +1030,1006 @@ int crush_do_rule(const struct crush_map *map,
...
+
+	const unsigned result_len;
+
+	const struct crush_work *crush_work;
+
+	// int[end_stepno - start_stepno][result_len]
+	int **step_vecs;
+};
+
+/// encapsulates output space, invariant through crush_msr_do_rule invocation
```
Comment (athanatos) [inline src/crush/mapper.c:1064]: invariant isn't the right word -- the output space remains the same but is mutated as items are output (out[i] and returned_so_far).  TODO

**Review by ljflores — APPROVED** [review]

**Comment by athanatos** [comment]
jenkins test make check arm64

**Comment by anthonyeleven** [comment]
Suggest a few sentences in the PR description.

**Comment by ljflores** [comment]
Rados-approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomctbflL3Ja1936-wip-yuri3-testing-2024-01-22-1155

**Comment by athanatos** [comment]
Let's hold of on merging for now, I'll take care of merging it.

**Comment by athanatos** [comment]
Once the revert merges, I'll repost this PR.

**Comment by athanatos** [comment]
https://github.com/ceph/ceph/pull/55332 is the new iteration of this PR.

**Pull request merged**